src/core/net/netif.cpp

```diff
@@ -535,5 +544,98 @@ bool Netif::IsUnicastAddress(const Address &aAddress) const
     return rval;
 }
 
+otError Netif::AddInternalUnicastAddress(const NetifUnicastAddress &aAddress)
+{
+    otError              error = OT_ERROR_NONE;
+    NetifUnicastAddress *entry;
+    const size_t         num = OT_ARRAY_LENGTH(mDynamicUnicasts);
+
+    VerifyOrExit(!aAddress.GetAddress().IsLinkLocal(), error = OT_ERROR_INVALID_ARGS);
+
+    for (entry = mUnicastAddresses; entry; entry = entry->GetNext())
+    {
+        if (entry->GetAddress() == aAddress.GetAddress())
+        {
+            VerifyOrExit(entry >= &mDynamicUnicasts[0] && entry < &mDynamicUnicasts[mInternalDynamicUnicastsNumber],
+                         error = OT_ERROR_INVALID_ARGS);
+
+            entry->mPrefixLength = aAddress.mPrefixLength;
+            entry->mPreferred    = aAddress.mPreferred;
+            entry->mValid        = aAddress.mValid;
+            ExitNow();
+        }
+    }
+
+    VerifyOrExit(mInternalDynamicUnicastsNumber + mExternalDynamicUnicastsNumber < num, error = OT_ERROR_NO_BUFS);
+
+    entry = &mDynamicUnicasts[mInternalDynamicUnicastsNumber++];
```

Comment:
How about using the `mUnicastAddressPool` instead of `mDynamicUnicasts`? 

Mainly to avoid confusion with static and dynamic IP address networking [terminology](https://support.google.com/fiber/answer/3547208?hl=en).